### PR TITLE
Enhance Loop Unroller

### PR DESCRIPTION
**Pull request by mostafaelhoushi**
We have noticed that the unrolling factor passed to `opt` as `--unroll-count=<num>` cli option doesn't always take effect.
This PR provides a custom LLVM loop unroller that enforces the unroll count factor provided in an action.. 

To run custom unroller:
```
bazel run //examples/example_unrolling_service/loop_unroller:loop_unroller -- <input>.ll -o <output>.ll  --funroll-count=<num>
```

To run end-to-end example:
```
bazel run //examples/example_unrolling_service:example
```

### CODE REVIEW

File: examples/example_unrolling_service/loop_unroller/loop_unroller.cc [hunk 1]
```diff
@@ -0,0 +1,147 @@
+#include <algorithm>
```
Comment (ChrisCummins) [inline examples/example_unrolling_service/loop_unroller/loop_unroller.cc:5]: This file needs a license header

File: examples/example_unrolling_service/loop_unroller/loop_unroller.cc [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/SourceMgr.h"
+
+using namespace llvm;
+#define DEBUG_TYPE "LoopUnroller"
+
+#define METHOD 1
+
+#if (METHOD == 1)
+// obtained from https://stackoverflow.com/a/33565910/3880948
```
Comment (ChrisCummins) [inline examples/example_unrolling_service/loop_unroller/loop_unroller.cc]: [LGTM](https://github.com/facebookresearch/CompilerGym/blob/development/compiler_gym/third_party/llvm/InstCount.cc#L13) :)

**Comment by mostafaelhoushi** [comment]
Update: Thanks for @hughleat  for his help. Counting the loops is now solved. Next step is to update the loop metadata

File: examples/example_unrolling_service/loop_unroller/loop_unroller.cc [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+/// Input LLVM module file name.
+cl::opt<std::string> InputFilename(cl::Positional, cl::desc("Specify input filename"),
+                                   cl::value_desc("filename"), cl::init("-"));
+/// Output LLVM module file name.
+cl::opt<std::string> OutputFilename("o", cl::desc("Specify output filename"),
+                                    cl::value_desc("filename"), cl::init("-"));
+
+namespace llvm {
+// The INITIALIZE_PASS_XXX macros put the initialiser in the llvm namespace.
```
Comment (ChrisCummins) [inline examples/example_unrolling_service/loop_unroller/loop_unroller.cc:68]: Total bikeshedding, but what about putting the whole file (except `main`) inside `llvm::`?

File: examples/example_unrolling_service/loop_unroller/loop_unroller.cc [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+  virtual void getAnalysisUsage(AnalysisUsage& AU) const override {
+    AU.addRequired<LoopInfoWrapperPass>();
+  }
+
+  bool runOnFunction(llvm::Function& F) override {
+    LoopInfo& LI = getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
+    auto Loops = LI.getLoopsInPreorder();
+
+    // Should reall account for module, too.
```
Comment (ChrisCummins) [inline examples/example_unrolling_service/loop_unroller/loop_unroller.cc]: Typo "reall"

**Review by ChrisCummins — COMMENTED** [review]
I had a pick through the code and it all looks good to me (minor comments inline). I'm guessing with this being WIP that the glue code to interface with an environment is coming later. Just ping me whenever you want me to take a look through :-)

Cheers,
Chris

File: examples/example_unrolling_service/loop_unroller/loop_unroller.cc (hunk 1 above)
Comment (mostafaelhoushi) [reply examples/example_unrolling_service/loop_unroller/loop_unroller.cc:5]: Done

File: examples/example_unrolling_service/loop_unroller/loop_unroller.cc (hunk 3 above)
Comment (mostafaelhoushi) [reply examples/example_unrolling_service/loop_unroller/loop_unroller.cc:68]: Done

**Comment by mostafaelhoushi** [comment]
> I had a pick through the code and it all looks good to me (minor comments inline). I'm guessing with this being WIP that the glue code to interface with an environment is coming later. Just ping me whenever you want me to take a look through :-)
> 
> Cheers, Chris

Thanks Chris. I have addressed all the comments.
Yes, now I have updated the environment to call the custom unroller. I have also made it optional in case someone wants to use LLVM's default unroller

**Comment by mostafaelhoushi** [comment]
> Just ping me whenever you want me to take a look through :-)

I merged it without pinging you :( 
Do you want to take another look and make changes?

**Comment by ChrisCummins** [comment]
> I merged it without pinging you :( Do you want to take another look and make changes?

No worries! Code looks great :) I'll try it out and let you know if i encounter any problems

Cheers,
Chris

**Pull request merged**